### ISSUE TITLE
sort by order ascending - Change Order of Review Cards in Regular Decks

### ISSUE THREAD

**Issue by ijgnd** [lovac42/HoochieMama#1]
Thank you for another wonderful add-on !! Also thanks for reading this long post ...

&nbsp;

"randomization of sub-decks on the v1 scheduler " in an add-on for 2.0 means to have the best part of the V2 scheduler (imo) while you can sync with AnkiDroid and use all the add-ons that haven't been ported yet. This seems like the most important new add-on of the last year to me. If my understanding is correct you could add this to the description so that more people find this gem.

&nbsp;

You warn against other add-ons that monkey patch "_fillRev". Very few people know this term so the warning doesn't help them. Luckily at the moment the only add-on from Ankiweb that monkey patches "_fillRev" is [Change Order of Review Cards in Regular Decks](https://ankiweb.net/shared/info/3731265543). Maybe you could explicitly mention this add-on.

&nbsp;

"Change Order of Review Cards in Regular Decks" is one of my favorite add-ons: If I have an interval of one day and I review a day late that makes a huge difference. If I have an interval of 127 days it makes no difference to review after 133 days. It's unfortunate that now I have to chose. For me "Change Order ..." hasn't noticeably slowed down Anki on the two computers I used it on. It would be great if your add-on also had an option to sort by increasing intervals. I know that it's brazen to ask for such a massive new feature but I'm afraid that my code will fail and/or have unintended consequences. Also if there were a built-in version I'm sure many other people would profit.

So far I have this: To make it easier for me I removed the serenity part. I replaced your lines 36-67 with

```python
    if lim:
        self._revQueue = self.col.db.list("""
select id from cards where
did in %s and queue = 2 and due <= ?
order by ivl asc
limit ?""" % (ids2str(self.col.decks.active())),
                self.today, lim)

        if self._revQueue:
            return True
```

Do you think that this would work? I'd be grateful for any hints but I fully understand if you ignore this post.

**Comment by lovac42** [comment]
Thanks I was trying to find the name of that addon, was it changed recently? I remember the old version didn't have a github link.

For the mods, I try not to deviate from the original too much. They wouldn't work well for everyone. But you can edit the SQL to rebuilt the queue for your personal needs at the cost of performance and other factors. Your code above looks fine, it may skew everything towards one deck, but if you set a large limit, it wouldn't be a problem.

**Comment by ijgnd** [comment]
Thanks for your hints about the code. I will just experiment. 

The add-on name hasn't changed but the ownership:

the add-on used to be developed by the the anki developer and was available at https://github.com/dae/ankiplugins.

This changed the ownership: https://github.com/dae/ankiplugins/commit/bb0ce363c7a93f17ed257e34dfe073118c4cecfd

the [new repo](https://github.com/xquercus/change-order-of-review-cards) is linked from ankiweb .

**Comment by ijgnd** [comment]
Thank you!! Your help is amazing.

I have seen that you quote me in the description of your add-on. When I wrote I didn't mean that you quote me verbatim but just mention that it works with Ankidroid and old add-ons. Though upon re-reading the [list of changes in the scheduler for 2.1](https://anki.tenderapp.com/kb/anki-ecosystem/experiment-scheduling-changes-in-anki-21) I still agree with that sentence.

I will test out over the coming days and report back. 

About ascending intervals in general: I wonder why you had "1 or 2 towards the end". This could be because you remember very mature cards less well and/or because cards that are shown later are remembered less well because your exhausted/tired etc. 

If the problem is about the cards themselves you might be in a similar situation as me [1]. 

But I'm afraid that this is also the latter. Until now I have neglected this point. This would be a very good reason not to use "ascending intervals". But still I regularly don't finish all the cards that are due in a day and I still think that my comparison (1 day interval and one day late vs. 127 days interval and six days late) is useful.

Maybe the ideal order would look like this:?
- cards with an interval shorter than some threshold like 10 days in ascending order
- rest of the cards due today (in random order)
- rest of the cards (in random order or maybe even in descending order?)





----

[1] I hate forgetting/rescheduling cards with very long intervals. It feels like I wasted reviews (even if I don't use a lapse interval of 0%). So I even [made an add-on for myself](https://github.com/ijgnd/Anki20__scheduler_mods/blob/master/shorter_ivls_for_mature_cards/Reduce_Interval_growth_for_very_mature_cards_alpha.py) to reduce the next interval for very mature cards which slightly increases the retention rate. Lowering the intervals overall (with a reduced interval modifier) is not an option because this would lead to a lot of additional work (see the [calculations in the manual](https://apps.ankiweb.net/docs/manual20.html#reviews)). I'm not sure if this add-on makes sense if psychologically you can better handle forgetting a very mature card. Maybe the additional workload could be justified like this: If each review lasts 10 seconds and ... 

- if the next review interval will be 500 days then the "time investment needed to remember the card contents per day" is 10/500 = 0,02 seconds/day (for the next 500 days). 
- if the next review interval is 20 days then the "time investment needed to remember the card contents per day" is 10/20 = 0,5 seconds/day (for the next 20 days). 

btw: Does this make sense to you? I am posting this here only because I haven't got any feedback on this idea yet.

**Comment by ijgnd** [comment]
Thanks for your reply and the quick release of a new version !!

At the moment the order is wrong for me: I see overdue cards (cards that were due before today) first, and only then I see cards due today. I think you want the opposite. As far as I see Anki uses pop() to get the next card, so I changed [line 92](https://github.com/lovac42/HoochieMama/blob/master/src/hoochieMama.py#L92) `if arr: self._revQueue.extend(arr)` to ` if arr: self._revQueue = arr + self._revQueue`. Now it seems to work.

&nbsp;

I tried to get the behavior of "Change Order of Cards ..." into your code, i.e. not prioritize today's reviews. So I made an additional if-loop. https://github.com/ijgnd/HoochieMama/commit/f0c7c90f2d520460d2a04d42b1c6afd9f62935ce   Would this work?   

&nbsp;
I just saw you released another clever add-on (Fuzz). Thank's so much.

**Comment by lovac42** [comment]
Your code looks fine, but I've found some sync errors in my code so I updated them and took the liberty of adding some of your ideas to my next version. Funny how no one reported the sync errors :-/
https://github.com/lovac42/HoochieMama/blob/deque/src/hoochieMama.py

Since you used a queue size of 1024 (way too big), I added a deque to improve the performance. If that runs into any compatibility issues with other addons let me know ;-) But I don't think there are any published addons that manipulates sched._revQueue directly.
